e2e/README.md

```diff
@@ -145,7 +149,1089 @@ func E2ETests(env e2e.TestEnv) testhelper.Tests {
+    at once, they could interfere with each other.
+  - Examples where this concern arises include any test that would potentially
+    change, or be sensitive to, the contents of files inside the user's
+    `$HOME/.singularity` directory, such as `remote.yaml`, `docker-config.json`,
+    and others, as well as any test that changes the content of the system
+    `singularity.conf`.
+
+To deal with such cases, e2e/internal/testhelper/testhelper.go defines a
+function `testhelper.NoParallel(func(*testing.T)) func(*testing.T)`. This
+function returns the argument it was given, but first adds it to a data
+structure that holds all the test functions that *cannot* be run in parallel.
+The suite.Run() function in e2e/internal/testhelper/testhelper.go then makes
+sure that these tests are run sequentially, and not in parallel with any other
+tests.
+
+It is for this reason that a typical test name in the e2e suite looks as follows:
+
+```plain
+<...>
+TestE2E/PAR/BUILD/build_with_bind_mount
+<...>
+TestE2E/SEQ/DOCKER/cred_prio
+<...>
+```
+
+`TestE2E` is the test name for the entire e2e suite; it is followed by `PAR`,
+for the set of tests run in parallel, or by `SEQ`, for those tests that cannot
+be run in parallel and are run sequentially.
+
+The fact that the testhelper.NoParallel() function returns its argument as its
+sole return value makes it handy to use in the construction of
+`testhelper.Tests` maps. Here, for example, is the E2ETests() function of the
+"REMOTE" tests group (note that `testhelper.NoParallel` is assigned to the local
+variable `np` for the sake of brevity, another best practice in writing
+E2ETests() functions):
+
+```go
+func E2ETests(env e2e.TestEnv) testhelper.Tests {
+	c := ctx{
+		env: env,
+	}
+
+	np := testhelper.NoParallel
+
+	return testhelper.Tests{
+		"add":            c.remoteAdd,
+		"list":           c.remoteList,
+		"default or not": c.remoteDefaultOrNot,
+		"remove":         c.remoteRemove,
+		"status":         c.remoteStatus,
+		"test help":      c.remoteTestHelp,
+		"use":            c.remoteUse,
+		"use exclusive":  np(c.remoteUseExclusive),
+	}
+}
+```
+
+As can be seen here, the c.remoteUseExclusive() test cannot be run in parallel,
+but it can be marked for sequential running *and* added to the
+`testhelper.Tests` map in one fell swoop, by making use of the return value of
+testhelper.NoParallel().
+
+## Useful utility functions
+
+Some useful utility functions for e2e testing have already been discussed above,
+including:
+
+- [`e2e.WriteTempFile(dir, pattern, content string) (string,
+  error)`](#temporary-dirs--files-and-cleanup)
+- [`e2e.MakeTempDir(t *testing.T, baseDir string, prefix string, context string)
+  (string, func(t *testing.T))`](#temporary-dirs--files-and-cleanup)
+
+Here are some additional utility functions that are available, and which are
+particularly useful for writing e2e tests:
+
+- The `require` package (internal/pkg/test/tool/require/require.go)
```

Comment:
Ideally add a link to the file in GitHub main branch.

Reply:
Ooh good idea!